Substitute memset/memcpy calls with faster alternatives

qwe661234 opened the pull request:
1. Because RISC-V use jal instruction as function call, so we remove recursive jump translation to detect function call.
2. To identify memset/memcpy, we analyze the pattern of the first basic block of memset/memcpy for comparison. If the basic block aligns with the pattern, we extract the instruction sequence starting from the initial PC of the basic block and then compare it with the pre-recorded memset/memcpy instruction sequence. Upon detecting a memset/memcpy, we substitute the basic block with the standard library function memset/memcpy.

Based on performance results below, we gain 2% performance improvement when running coreMark and dhrysone. The performance impact is resulted in the number of invoking memset/memcpy times.

* Intel Core i7-11700

|   Metric   | origin  | proposed |Speedup|
|------------| ---------| ---------| -------|
| Dhrystone  | 1413.11 | 1447.11  |  +2%  |
| CoreMark   | 1497.35 | 1525.31  |  +2%  |

See: #153

Commit:
Remove recursive jump translation

jserv commented on src/emulate.c:
```diff
@@ -593,18 +597,57 @@ static void block_translate(riscv_t *rv, block_t *block)
 
+bool parse_memset(riscv_t *rv)
+{
+    uint32_t memset_insn[] = {
```
Append the corresponding assembly as inlined comments for reference purpose.

jserv replied:
Add `static const` qualifier for `memset_insn`.

jserv commented on src/emulate.c:
```diff
@@ -593,18 +597,57 @@ static void block_translate(riscv_t *rv, block_t *block)
         ir->tailcall = next_ir->tailcall;                                  \
     }
 
+bool parse_memset(riscv_t *rv)
```
Add `static` qualifier.

jserv commented on src/emulate.c:
```diff
@@ -593,18 +597,57 @@ static void block_translate(riscv_t *rv, block_t *block)
+        0xfa0680e7, 0x00028093, 0xff078793, 0x40f70733, 0x00f60633, 0xf6c378e3,
+        0xf3dff06f};
+    uint32_t tmp_pc = rv->PC;
+    for (int i = 0; i < 55; i++) {
```
`memcmp` provided by glibc or other C runtimes is usually optimized for modern microprocessors, and it can be much faster than simple iterations. Do you have any idea to utilize `memcmp`?

jserv reviewed (changes requested):
Add memcpy substitution as well. Applications such as Doom can benefit.

jserv commented on src/emulate.c:
```diff
@@ -593,18 +597,57 @@ static void block_translate(riscv_t *rv, block_t *block)
         int32_t count = 0, sign = 1;
         switch (ir->opcode) {
+        case rv_insn_addi:
+            if (ir->imm == 15 && ir->rd == 6 && ir->rs1 == 0) {
```
Add more comments.

jserv commented on src/emulate.c:
```diff
@@ -593,18 +613,255 @@ static void block_translate(riscv_t *rv, block_t *block)
         ir->tailcall = next_ir->tailcall;                                  \
     }
 
+static bool parse_memset(riscv_t *rv)
```
The function name should be `detect_memset`.

jserv commented on src/emulate.c:
```diff
@@ -593,18 +613,255 @@ static void block_translate(riscv_t *rv, block_t *block)
+    return true;
+}
+
+static bool parse_memcpy(riscv_t *rv)
```
Ditto. Should be `detect_memcpy`.

jserv commented:
If a certain region is recognized as containing a sequence of known `memset` or `memcpy` instructions, is it possible to replace the invocation of such a region? In other words, instead of replacing the region with a faster alternative, can we directly modify the instructions on the caller's side?

jserv commented:
Rather than identifying `memcpy` / `memset` during the execution phase, I am considering whether it might be more feasible to detect them during the decoding phase. This approach would involve examining the initial portion of the instruction sequence in the buffer and using `memcmp` to validate, resulting in faster and deterministic instruction substitution.

jserv commented on src/emulate.c:
```diff
@@ -593,18 +613,314 @@ static void block_translate(riscv_t *rv, block_t *block)
+    return true;
+}
+
+static bool parse_memset2(riscv_t *rv)
```
Should be `detect_memset2`. Of course, it would be better if one routine can manipulate several patterns.

jserv commented on src/emulate.c:
```diff
@@ -593,18 +613,314 @@ static void block_translate(riscv_t *rv, block_t *block)
+        0x00008067, /* ret */
+    };
+    uint32_t tmp_pc = rv->PC;
+    for (int i = 0; i < 7; i++) {
```
Replace the hardcoded `7` with `ARRAYS_SIZE` macro.

jserv commented on src/emulate.c:
```diff
@@ -593,18 +613,314 @@ static void block_translate(riscv_t *rv, block_t *block)
+        0xfcdff06f, /* j	0x21520 */
+    };
+    uint32_t tmp_pc = rv->PC;
+    for (int i = 0; i < 105; i++) {
```
Ditto. Utilize `ARRAYS_SIZE` macro.

jserv commented on src/emulate.c:
```diff
@@ -35,6 +35,8 @@ extern struct target_ops gdbstub_ops;
 #include "state.h"
 #include "utils.h"
 
+#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof((arr)[0]))
```
Don't do that! The macro `ARRAYS_SIZE` is already defined in `src/common.h`.

ChinYikMing commented:
To maintain the consistency of accessing riscv's register, I think the code below which `X[1]`, `X[10]`, `X[11]` and `X[12]` can be rewrite with `X[rv_reg_ra]`, `X[rv_reg_a0]`, `X[rv_reg_a1]` and `X[rv_reg_a2]` respectively.

```c
memset((char *) m->mem_base + rv->X[10], rv->X[11], rv->X[12]);
rv->PC = rv->X[1] & ~1U;
```

Commit:
Inject memset

1. Because RISC-V use jal instruction as function call, so we remove recursive
jump tranlation to detect function call.
2. To identify memset, we analyze the pattern of the first basic block of
memset for comparison. If the basic block aligns with the pattern, we extract
the instruction sequence starting from the initial PC of the basic block
and then compare it with the pre-recorded memset instruction sequence.
Upon detecting a memset, we substitute the basic block with the standard
library function memset.

Based on performance results below, we gain 3% performance improvement when
running coreMark and lost 1% performance when running dhrysone. The performance
impact is resulted in the number of invoking memset times.

| Metric     | origin  | proposed |Speedup|memset times|
|------------+---------+----------+-------+------------|
| Dhrystone  | 2293.25 | 2267.00  |  -1%  | 4          |
| CoreMark   | 1881.41 | 1933.81  |  +3%  | 3124365    |
